### PR TITLE
Wrap RET: batch 7 (Transfers)

### PR DESCRIPTION
**Pull request by CyonAlexRDX**
Add two variants of representation of AssetsTransfers, one which is **per recipient** (which is how transfers are build in UI in wallets) and one which is **per resource** (which is the format Manifest Building always uses):
* PerRecipientAssetTransfers
* PerResourceAssetTransfers

A `transpose()` method maps from `PerRecipientAssetTransfers` -> `PerResourceAssetTransfers`.

And implementation of TransactionManifest building using  `PerResourceAssetTransfers`.

Providing both `manifest_per_asset_transfers` and `manifest_per_recipient_transfers` to wallets makes it easier for wallets during integration to use Sargon. Maybe iOS and Android differs and having both option makes it easier to integrate.

### CODE REVIEW

File: src/wrapped_radix_engine_toolkit/high_level/ret_api.rs [hunk 1]
```diff
@@ -145,29 +145,27 @@ pub fn debug_print_compiled_notarized_intent(
...
-// ) -> Result<bool> {
-//     unreachable!()
-// }
+/// Uses `per_asset_transfers` after having transposed the `PerRecipientAssetTransfers`
+/// into `PerAssetTransfers`. We always use `PerAssetTransfers` when building the manifest
+/// since it is more efficient (allows a single withdraw per resource) => fewer instruction =>
+/// cheaper TX fee for user.
+#[uniffi::export]
+pub fn manifest_per_recipient_transfers(
+    transfers: PerRecipientAssetTransfers,
```
Comment (CyonAlexRDX) [inline src/wrapped_radix_engine_toolkit/high_level/ret_api.rs:154]: wallets need to specify which resources for which recipients require `deposit` vs `try_deposit_or_abort`

Comment (CyonAlexRDX) [inline src/wrapped_radix_engine_toolkit/high_level/ret_api.rs:154]: wallets need to specify which resources for which recipients require `deposit` vs `try_deposit_or_abort`, part of `PerRecipientAssetTransfers`

Comment (CyonAlexRDX) [inline src/wrapped_radix_engine_toolkit/high_level/ret_api.rs:154]: wallets need to specify which resources for which recipients require `deposit` vs `try_deposit_or_abort`, part of `PerRecipientAssetTransfers`

File: src/wrapped_radix_engine_toolkit/high_level/ret_api.rs [hunk 2]
```diff
@@ -145,29 +145,27 @@ pub fn debug_print_compiled_notarized_intent(
...
+) -> TransactionManifest {
+    TransactionManifest::per_recipient_transfers(transfers)
+}
+
+#[uniffi::export]
+pub fn manifest_per_asset_transfers(
+    transfers: PerAssetTransfers,
```
Comment (CyonAlexRDX) [inline src/wrapped_radix_engine_toolkit/high_level/ret_api.rs:161]: wallets need to specify which resources for which recipients require `deposit` vs `try_deposit_or_abort`, part of `PerAssetTransfers `

File: src/wrapped_radix_engine_toolkit/high_level/sargon_specific_types/assets_transfers/per_asset/per_asset_fungible_resource.rs [hunk 3]
```diff
@@ -0,0 +1,85 @@
+use crate::prelude::*;
+
+#[derive(Clone, Debug, PartialEq, Eq, Hash, uniffi::Record)]
+pub struct PerAssetFungibleResource {
+    pub resource_address: ResourceAddress,
+    pub divisibility: Option<i32>,
```
Comment (CyonAlexRDX) [inline src/wrapped_radix_engine_toolkit/high_level/sargon_specific_types/assets_transfers/per_asset/per_asset_fungible_resource.rs:6]: used for rounding, potentially we can require wallets to already have rounded, but I think this is safer.

File: src/wrapped_radix_engine_toolkit/high_level/sargon_specific_types/assets_transfers/per_recipient/per_recipient_asset_transfers.rs [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+
+    fn sample_other() -> Self {
+        Self::sample_stokenet_other()
+    }
+}
+
+impl PerRecipientAssetTransfers {
+    /// Transpose: `PerRecipient` -> `PerAsset`
+    pub fn transpose(&self) -> PerAssetTransfers {
+        let mut per_asset_fungibles = IndexMap::<
```
Comment (CyonAlexRDX) [inline src/wrapped_radix_engine_toolkit/high_level/sargon_specific_types/assets_transfers/per_recipient/per_recipient_asset_transfers.rs:71]: group by `ResourceAddress` for fungibles and non fungibles.

File: src/wrapped_radix_engine_toolkit/high_level/sargon_specific_types/assets_transfers/per_recipient/per_recipient_asset_transfers.rs [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+                        ),
+                    );
+                }
+            });
+
+            x.non_fungibles.clone().into_iter().for_each(|y| {
+                if let Some(existing_transfers) =
+                    per_asset_non_fungibles.get_mut(&y.resource_address)
+                {
+                    existing_transfers.expanded((recipient, y.clone()));
```
Comment (CyonAlexRDX) [inline src/wrapped_radix_engine_toolkit/high_level/sargon_specific_types/assets_transfers/per_recipient/per_recipient_asset_transfers.rs:111]: this is the second transfer of that resource, expand the existing list of resources

Comment (CyonAlexRDX) [inline src/wrapped_radix_engine_toolkit/high_level/sargon_specific_types/assets_transfers/per_recipient/per_recipient_asset_transfers.rs:111]: this is the second transfer of that resource, expand the existing list of transfers

File: src/wrapped_radix_engine_toolkit/high_level/sargon_specific_types/assets_transfers/per_recipient/per_recipient_asset_transfers.rs [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+                } else {
+                    per_asset_non_fungibles.insert(
+                        y.resource_address.clone(),
+                        PerAssetTransfersOfNonFungibleResource::new(
```
Comment (CyonAlexRDX) [inline src/wrapped_radix_engine_toolkit/high_level/sargon_specific_types/assets_transfers/per_recipient/per_recipient_asset_transfers.rs:115]: first transfer of this NF resource

Comment (CyonAlexRDX) [inline src/wrapped_radix_engine_toolkit/high_level/sargon_specific_types/assets_transfers/per_recipient/per_recipient_asset_transfers.rs:115]: first transfer of this NF resource

File: src/wrapped_radix_engine_toolkit/high_level/sargon_specific_types/assets_transfers/per_recipient/per_recipient_asset_transfers.rs [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+
+            x.fungibles.clone().into_iter().for_each(|y| {
+                if let Some(existing_transfers) =
+                    per_asset_fungibles.get_mut(&y.resource_address)
+                {
+                    existing_transfers.expanded((recipient, y.clone()));
+                } else {
+                    per_asset_fungibles.insert(
+                        y.resource_address.clone(),
+                        PerAssetTransfersOfFungibleResource::new(
```
Comment (CyonAlexRDX) [inline src/wrapped_radix_engine_toolkit/high_level/sargon_specific_types/assets_transfers/per_recipient/per_recipient_asset_transfers.rs:93]: first transfer of this fungible resource

File: src/wrapped_radix_engine_toolkit/high_level/sargon_specific_types/assets_transfers/per_recipient/per_recipient_asset_transfers.rs [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+                SUT::sample_stokenet_other(),
+            ])
+            .len(),
+            4
+        )
+    }
+
+    // #[test]
+    // fn transpose() {
+    //     pretty_assertions::assert_eq!(SUT::sample().transpose(), PerAssetTransfers::sample())
```
Comment (CyonAlexRDX) [inline src/wrapped_radix_engine_toolkit/high_level/sargon_specific_types/assets_transfers/per_recipient/per_recipient_asset_transfers.rs:174]: can add more transpose tests maybe, but I'm already at 100% coverage actually.

File: src/wrapped_radix_engine_toolkit/high_level/sargon_specific_types/assets_transfers/per_recipient/per_recipient_non_fungibles_transfer.rs [hunk 9]
```diff
@@ -0,0 +1,109 @@
+use crate::prelude::*;
+
+impl PerRecipientNonFungiblesTransfer {
```
Comment (CyonAlexRDX) [inline src/wrapped_radix_engine_toolkit/high_level/sargon_specific_types/assets_transfers/per_recipient/per_recipient_non_fungibles_transfer.rs:3]: The `PerRecipientNonFungiblesTransfer ` type is declared with macro in `transfer_types.rs`

File: src/wrapped_radix_engine_toolkit/high_level/sargon_specific_types/assets_transfers/per_recipient/per_recipient_fungible_transfer.rs [hunk 10]
```diff
@@ -0,0 +1,83 @@
+use crate::prelude::*;
+
+impl PerRecipientFungibleTransfer {
```
Comment (CyonAlexRDX) [inline src/wrapped_radix_engine_toolkit/high_level/sargon_specific_types/assets_transfers/per_recipient/per_recipient_fungible_transfer.rs:3]: The `PerRecipientFungibleTransfer ` type is declared with macro in `transfer_types.rs`

File: src/wrapped_radix_engine_toolkit/high_level/sargon_specific_types/assets_transfers/transfer_types.rs [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+                pub(crate) use_try_deposit_or_abort: bool,
+
+                $($fields)*
+
+            }
+        }
+    };
+}
+
+macro_rules! decl_per_asset_transfer_of {
```
Comment (CyonAlexRDX) [inline src/wrapped_radix_engine_toolkit/high_level/sargon_specific_types/assets_transfers/transfer_types.rs:33]: a macro using a macro 🤯

File: src/wrapped_radix_engine_toolkit/high_level/sargon_specific_types/assets_transfers/transfer_types.rs [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+                        return builder
+                            .deposit(self.recipient.account_address(), bucket);
+                    }
+                }
+            }
+        }
+    };
+}
+
+macro_rules! decl_per_recipient_transfer_of {
```
Comment (CyonAlexRDX) [inline src/wrapped_radix_engine_toolkit/high_level/sargon_specific_types/assets_transfers/transfer_types.rs:74]: a macro using a macro 🤯

File: src/wrapped_radix_engine_toolkit/high_level/sargon_specific_types/assets_transfers/transfer_types.rs [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+            )*
+            #[derive(Clone, Debug, PartialEq, Eq, Hash, uniffi::Record)]
+            pub struct [< $struct_prefix $struct_name Transfer >] {
+
+                /// If `true` the `try_deposit_batch_or_abort` method will be used instead of `deposit`,
+                /// typically wallets sets this to try if and only if the recipient is a self-owned account
+                /// (`AssetsTransfersRecipient::MyOwnAccount`) controlled by a DeviceFactorSource thy have
+                /// access to and which third party deposit setting's `DepositRule` is `AcceptKnown` and
+                /// which resource is known (`resource_address` is owned or has been owned before).
+                pub(crate) use_try_deposit_or_abort: bool,
```
Comment (CyonAlexRDX) [inline src/wrapped_radix_engine_toolkit/high_level/sargon_specific_types/assets_transfers/transfer_types.rs:24]: Thanks to macros we can have great documentation of this property in one place.

File: src/wrapped_radix_engine_toolkit/high_level/sargon_specific_types/assets_transfers/per_asset/per_asset_non_fungible_transfer.rs [hunk 14]
```diff
@@ -0,0 +1,136 @@
+use crate::prelude::*;
+
+impl PerAssetNonFungibleTransfer {
```
Comment (CyonAlexRDX) [inline src/wrapped_radix_engine_toolkit/high_level/sargon_specific_types/assets_transfers/per_asset/per_asset_non_fungible_transfer.rs:3]: The `PerAssetNonFungibleTransfer ` type is declared with macro in `transfer_types.rs`

File: src/wrapped_radix_engine_toolkit/high_level/sargon_specific_types/assets_transfers/per_asset/per_asset_fungible_transfer.rs [hunk 15]
```diff
@@ -0,0 +1,123 @@
+use crate::prelude::*;
+
+impl PerAssetFungibleTransfer {
```
Comment (CyonAlexRDX) [inline src/wrapped_radix_engine_toolkit/high_level/sargon_specific_types/assets_transfers/per_asset/per_asset_fungible_transfer.rs:3]: The `PerAssetFungibleTransfer ` type is declared with macro in `transfer_types.rs`

**Review by GhenadieVP — APPROVED** [review]
Awesome!

**Pull request merged**